File: Cattaneo-2023/CFPT_2024_wp-main/CFPT_2022_wp.R
###############################################################################
## Replication File for Cattaneo, Feng, Palomba, and Titiunik (2022)
###############################################################################


######################################################################################
# Three predictands: unit-time, time, and unit
######################################################################################

# Load stuff 

pacman::p_load(scpi, haven)
path <- 'YOUR_PATH_HERE'

sims <- 200

path.data <- paste0(path,"data/")
path.fig <- paste0(path,"fig/")

set.seed(8894)

data <- haven::read_dta(paste0(path.data, "final_data.dta"))
data <- subset(data, restricted2 == 1)
data$lgdp <- log(data$rgdppp)
post.est <- 10

units <- df <- est <- res <- list()
df.unit <- est.unit <- res.unit <- list()
df.time <- est.time <- res.time <- list()

features <- list(c("lgdp","lsc"))
covs.adj <- list("lgdp" = c("constant","trend"), "lsc" = c("constant","trend"))

for (cont in c("Africa", "Asia", "Europe", "South America", "North America")) {
  print(paste0("Currently working on ", cont))
  
  print("Estimating whole treatment effect time series")
  
  units[[cont]] <- unique(subset(data, continent == cont & treated == 1 & trDate <= 1992)$countryname)
  if (cont == "Europe") {
    units[[cont]] <- units[[cont]][!(units[[cont]] %in% c("Slovenia"))]
  }
  
  df[[cont]] <- scdataMulti(data, id.var = "countryname", outcome.var = "lgdp", 
                            treatment.var = "liberalization", time.var = "year", constant = TRUE, 
                            cointegrated.data = TRUE, post.est = post.est, units.est = units[[cont]],
                            features=features, cov.adj = covs.adj, anticipation = 1)
  
  res[[cont]] <- scpi(df[[cont]], sims = sims, cores = cores, w.constr = list("name" = "L1-L2"),
                      u.order = 1, u.lags = 1)
  
  # Analysis on average treatment
  print("Estimating average unit treatment effect")
  
  df.unit[[cont]] <- scdataMulti(data, id.var = "countryname", outcome.var = "lgdp", effect = "unit",
                                 treatment.var = "liberalization", time.var = "year", constant = TRUE, 
                                 cointegrated.data = TRUE, post.est = post.est, units.est = units[[cont]],
                                 features = features, cov.adj = covs.adj, anticipation = 1)
  
  res.unit[[cont]] <- scpi(df.unit[[cont]], sims = sims, cores = cores, w.constr = list("name" = "L1-L2"),
                           u.order = 1, u.lags = 1)
  
  
  # Analysis on average treatment on the treated
  print("Estimating average treatment effect")
  # if (cont == "Europe") {
  #   units[[cont]] <- unique(subset(data, continent == cont & treated == 1 & trDate <= 1992)$countryname)
  # }
  df.time[[cont]] <- scdataMulti(data, id.var = "countryname", outcome.var = "lgdp", effect = "time",
                                 treatment.var = "liberalization", time.var = "year", constant = T, 
                                 cointegrated.data = T, post.est = post.est, units.est = units[[cont]],
                                 features=features, cov.adj = covs.adj, anticipation = 1)
  
  res.time[[cont]] <- scpi(df.time[[cont]], sims = sims, cores = cores, w.constr = list("name" = "L1-L2"),
                           u.order = 1, u.lags = 1)
  
}

save.image(file = paste0(path.fig, "workspace_ridge.Rdata"))


######################################################################################
# Fourth predictand - Case study
######################################################################################

data <- haven::read_dta(paste0(path.data, "final_data.dta"))
data <- subset(data, restricted2 == 1)
data$lgdp <- log(data$rgdppp)
post.est <- 10

# One Feature
eu.tr.91 <- unique(subset(data, trDate %in% c(1991) & continent == "Europe")$countryname)

data.co <- subset(data, !(countryname %in% eu.tr.91))
data.tr <- subset(data, countryname %in% eu.tr.91)

data.tr.agg <- aggregate(data.tr[c("lgdp", "liberalization", "lsc")], 
                         by=list(data.tr$year), FUN=mean, na.rm = TRUE)
names(data.tr.agg) <- c("year","lgdp", "liberalization", "lsc")
data.tr.agg$countryname <- "Average Unit"
data.tr.agg$liberalization <- 1*(data.tr.agg$liberalization > 0)

data.agg <- rbind(data.tr.agg, data.co[names(data.tr.agg)])

features <- list(c("lgdp"))
covs.adj <- list(c("constant","trend"))

df.agg <- scdataMulti(data.agg, id.var = "countryname", outcome.var = "lgdp",
                      treatment.var = "liberalization", time.var = "year", constant = FALSE, 
                      cointegrated.data = T, post.est = 10, units.est = "Average Unit", 
                      features=features, cov.adj=covs.adj) 

res.avg.l1l2 <- scpi(df.agg, sims = sims, cores = 1, w.constr = list("name" = "L1-L2"),
                     u.order = 1, u.lags = 1)

res.avg <- scpi(df.agg, sims = sims, cores = 1, w.constr = list("name" = "simplex"),
                u.order = 1, u.lags = 1)

save.image(file = paste0(path.fig, "workspace_casestudy_excluded.Rdata"))


# two features
features <- list(c("lgdp","lsc"))
covs.adj <- list(c("constant","trend"))

df.agg <- scdataMulti(data.agg, id.var = "countryname", outcome.var = "lgdp",
                      treatment.var = "liberalization", time.var = "year", constant = TRUE, 
                      cointegrated.data = T, post.est = 10, units.est = "Average Unit", 
                      features=features, cov.adj=covs.adj) 

res.avg.l1l2 <- scpi(df.agg, sims = sims, cores = 1, w.constr = list("name" = "L1-L2"),
                     u.order = 1, u.lags = 1)

res.avg <- scpi(df.agg, sims = sims, cores = 1, w.constr = list("name" = "simplex"),
                u.order = 1, u.lags = 1)

save.image(file = paste0(path.fig, "workspace_casestudy.Rdata"))

File: Cattaneo-2023/CFPT_2024_wp-main/CFPT_2022_wp.py
###################################################################
# Replication file for Cattaneo, Feng, Palomba, and Titiunik (2022)
###################################################################

########################################
# Load SCPI_PKG package
import pandas
import numpy
import random
import os
from warnings import filterwarnings
from plotnine import ggtitle, ggsave, theme, element_blank
from copy import deepcopy

from scpi_pkg.scdata import scdata
from scpi_pkg.scdataMulti import scdataMulti
from scpi_pkg.scest import scest
from scpi_pkg.scpi import scpi
from scpi_pkg.scplot import scplot
from scpi_pkg.scplotMulti import scplotMulti

os.chdir('YOUR_PATH_HERE')
filterwarnings('ignore')
sims = 200
cores = 1

########################################
# Load database
data = pandas.read_stata('final_data.dta')
data = data.loc[data['restricted2'] == 1]
data['lgdp'] = numpy.log(data['rgdppp'])
data['year'] = pandas.DatetimeIndex(data['year']).year  # from datetime64 to int

########################################
# Analyze separately each continent

for method in ["L1-L2"]:

    for cont in ["Asia", "Africa", "Europe", "North America", "South America"]:

        units = data.loc[(data['continent'] == cont) & (data['treated'] == 1) &
                         (data['trDate'] <= 1992)]['countryname'].unique().tolist()

        if cont == "Europe":
            units = [un for un in units if un != "Slovenia"]

        ###############################################
        # unit-time treatment effect
        ###############################################
        print(method + ' - ' + cont + ': unit-time')
        aux = scdataMulti(df=data,
                          id_var='countryname',
                          treatment_var='liberalization',
                          outcome_var='lgdp',
                          time_var='year',
                          features={'treated': ['lgdp', 'lsc']},
                          constant=True,
                          cointegrated_data=True,
                          cov_adj={'treated': ['constant', 'trend']},
                          post_est=10,
                          units_est=units,
                          effect='unit-time',
                          anticipation=1)

        res = scpi(aux, w_constr={'name': 'simplex'}, sims=sims,
                   u_order=1, u_lags=1, cores=cores)
        p = scplotMulti(res, ptype='series', scales='free_y', ncols=2)

        p = (p + theme(axis_text_y=element_blank(),
                       axis_ticks_major_y=element_blank()))
        plotname = cont.replace(' ', '_') + '_unittime_' + method + '.png'
        ggsave(filename=plotname, plot=p)

        ###############################################
        # average unit post-treatment effect
        ###############################################
        print(method + ' - ' + cont + ': unit')
        aux = scdataMulti(df=data,
                          id_var='countryname',
                          treatment_var='liberalization',
                          outcome_var='lgdp',
                          time_var='year',
                          features={'treated': ['lgdp', 'lsc']},
                          constant=True,
                          cointegrated_data=True,
                          cov_adj={'treated': ['constant', 'trend']},
                          post_est=10,
                          units_est=units,
                          effect='unit')

        res = scpi(aux, w_constr={'name': 'simplex'}, sims=sims,
                   u_order=1, u_lags=1, cores=cores)
        p = scplotMulti(res, ptype='series')
        plotname = cont.replace(' ', '_') + '_unit_' + method + '.png'
        ggsave(filename=plotname, plot=p)

        ###############################################
        # average post-treatment effect on the treated
        ###############################################
        print(method + ' - ' + cont + ': time')
        aux = scdataMulti(df=data,
                          id_var='countryname',
                          treatment_var='liberalization',
                          outcome_var='lgdp',
                          time_var='year',
                          features={'treated': ['lgdp', 'lsc']},
                          constant=True,
                          cointegrated_data=True,
                          cov_adj={'treated': ['constant', 'trend']},
                          post_est=10,
                          units_est=units,
                          effect='time')

        res = scpi(aux, w_constr={'name': 'simplex'}, sims=sims,
                   u_order=1, u_lags=1, cores=cores)

        p = scplotMulti(res, ptype='series')
        plotname = cont.replace(' ', '_') + '_time_' + method + '.png'
        ggsave(filename=plotname, plot=p)
